### ISSUE TITLE
Caching issue with `getSchema` function

### ISSUE THREAD

**Issue by Berger92** [B2o5T/graphql-eslint#1232]
### Issue workflow progress

_Progress of the issue based on the [Contributor Workflow](https://github.com/the-guild-org/Stack/blob/master/CONTRIBUTING.md#a-typical-contributor-workflow)_

- [ ] 1. The issue provides a reproduction available on GitHub, Stackblitz or CodeSandbox
     > Please make sure the graphql-eslint version under `package.json` matches yours.
- [ ] 2. A failing test has been provided
- [ ] 3. A local solution has been provided
- [ ] 4. A pull request is pending review

---

**Describe the bug**

Hi 👋 

I've had some issues with tests using eslint's nodejs API and I realized it is because the parser [uses](https://github.com/B2o5T/graphql-eslint/blob/65fdf0dcc39a92a39fe9bd838a0c739a5e095e92/packages/plugin/src/schema.ts#L22) a cached schema. This also seems to be the reason why the eslint service needs to be restarted in IDEs when there is a schema change and some rules, e.g `no-unreachable-types`, don't pick up the change.

I appreciate that caching the schema has performance implications, but without being able to reset the cache it can lead to unexpected behavior.

**To Reproduce**

Valid schema:
```graphql
type Foo {
  bar: String!
}

type Query {
  foo: Foo!
}
```

Invalid schema:
```graphql
type Foo {
  bar: String!
}

type Query {
  hello: String!
}
```

```typescript
const pathToValidSchema = '...'
const pathToInvalidSchema = '...'

let eslint = new ESLint({
    useEslintrc: false,
    baseConfig: {
        parser: "@graphql-eslint/eslint-plugin",
        plugins: ["@graphql-eslint"],
        parserOptions: {
            schema: pathToValidSchema,
        },
        rules: {
           "@graphql-eslint/no-unreachable-types": "error"
        }
    },
})

let results = await eslint.lintFiles(pathToValidSchema)

expect(results[0].messages).toHaveLength(0) // ✅ 

eslint = new ESLint({
    useEslintrc: false,
    baseConfig: {
        parser: "@graphql-eslint/eslint-plugin",
        plugins: ["@graphql-eslint"],
        parserOptions: {
            schema: pathToInvalidSchema,
        },
        rules: {
           "@graphql-eslint/no-unreachable-types": "error"
        }
    },
})

results = await eslint.lintFiles(pathToInvalidSchema). 

expect(results[0].messages).toHaveLength(1) // 🟥  <--- This will return 0 errors because it is still using the previous schema

```

**Expected behavior**

It should use the new `paserOptions.schema` and therefore return the correct number of errors.

**Environment:**

- OS: MacOS
- `@graphql-eslint/eslint-plugin`: 3.12.0
- Node.js: 16.16

**Resolution**

1) As many of the rules [have a test guard](https://github.com/B2o5T/graphql-eslint/blob/65fdf0dcc39a92a39fe9bd838a0c739a5e095e92/packages/plugin/src/rules/relay-edge-types.ts#L28-L32) for caching, `getSchema` function could have one too.
```typescript
if (process.env.NODE_ENV !== 'test' && schemaCache.has(schemaKey)) {
    return schemaCache.get(schemaKey);
  }
```

2) Caching could be based on content and not filename(s)

While option 1) is quicker to do, option 2) would also make sure that this plugin is working to its fullest in IDEs.

**Comment by B2o5T** [comment]
Hi ! Future release with cache invalidation could be tested in the following alpha version `3.13.0-alpha-20221024000319-1089aa0`

**Comment by B2o5T** [comment]
released in [`@graphql-eslint/eslint-plugin@3.13.0`](https://github.com/B2o5T/graphql-eslint/releases/tag/release-1667313303457)
